File: cpp/data_structures/two_chapter/isPrimeNumber.c
#include <iostream>

using namespace std;

class Bitmap {
private:
	int *F, N;
	int *T; int top;
protected:
	bool valid (int r) {
		return (0 <= r) && (r < top);
	}
	bool erased (int k) {
		return valid (F[k]) && !(T[ F[k] ] + 1 + k); //T[ F[k] ] = -1 - k;
	}
public:
	Bitmap (int n = 8) {
		N = n;
		F = new int[N];
		T = new int[N];
		top = 0;
	}
	~Bitmap() {
		delete [] F;
		delete [] T;
	}

	void set (int k) {
		if (test (k) )
			return;
		if (!erased (k))
			F[k] = top++;
		T[ F[k] ] = k;
	}
	void clear (int k) {
		if (test(k))
			T[ F[k] ] = -1 - k;
	}
	bool test (int k) {
		return valid (F[k]) && (k == T[ F[k] ]);
	}
};

double sqrt(const double &d, const double precision) {
	double k = d;
	while (1) {
		if (k * k > d - precision && k * k < d + precision)
			break;
		k = 0.5 * (k + d / k);	//newton algorithm
	}
	return k;
}

void eratosthenes(const int& n) {
	Bitmap B(n);
	B.set(0);
	B.set(1);
	for (int i = 2; i < sqrt(n, 1); ++i) 
		if (!B.test(i))
			for (int j = i * i; j < n; j += i) {
				B.set(j);
			}
	for (int i = 0; i < n; ++i)
		if (!B.test(i))
			cout << i << ", ";
	cout << endl;
}

void isPrimeNumber(const int& n) {
	eratosthenes(n);
}

int main(int argc, char** argv)
{
	int n;
	while(cout << "please input number:", cin >> n)
		isPrimeNumber(n);


	return 0;
}

File: go/data_structure/2_chapter/isPrimeNumber.go
package main

import (
	"fmt"
)

func main() {
	var n int
	for true {
		fmt.Print("please input number:")
		fmt.Scanln(&n)
		fmt.Println(isPrimeNumber(n))
	}
}

var array []bool

func isPrimeNumber(n int) bool {
	if n >= len(array) {
		eratosthenes(n + 1)
	}
	return !array[n]
}

func eratosthenes(n int) {
	if n < 2 {
		n = 2
	}
	array = make([]bool, n)
	// 0 and 1 is not prime number
	array[0] = true
	array[1] = true
	for i := 2; float64(i) < sqrt(float64(n), 1.0); i++ {
		if !array[i] {
			for j := i * i; j < n; j += i {
				array[j] = true
			}
		}
	}
}

func sqrt(n, precision float64) float64 {
	res := n
	for res*res <= n-precision || res*res >= n+precision {
		res = (n/res + res) / 2
	}
	return res
}
